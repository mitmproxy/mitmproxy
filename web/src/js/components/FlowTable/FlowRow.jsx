import React from 'react'
import PropTypes from 'prop-types'
import classnames from 'classnames'
import {defaultColumnNames} from './FlowColumns'
import { pure } from '../../utils'
import {getDisplayColumns} from './FlowTableHead'
import { connect } from 'react-redux'

FlowRow.propTypes = {
    onSelect: PropTypes.func.isRequired,
    flow: PropTypes.object.isRequired,
    highlighted: PropTypes.bool,
    selected: PropTypes.bool,
}

function FlowRow({ flow, selected, highlighted, onSelect, displayColumnNames }) {
    const className = classnames({
        'selected': selected,
        'highlighted': highlighted,
        'intercepted': flow.intercepted,
        'has-request': flow.request,
        'has-response': flow.response,
    })

    const displayColumns = getDisplayColumns(displayColumnNames)

    return (
        <tr className={className} onClick={() => onSelect(flow.id)}>
            {displayColumns.map(Column => (
                <Column key={Column.name} flow={flow}/>
            ))}
        </tr>
    )
}

export default connect(
    state => ({
        displayColumnNames: state.options["web_columns"] ? state.options["web_columns"].value : defaultColumnNames,
    })
)(pure(FlowRow))
