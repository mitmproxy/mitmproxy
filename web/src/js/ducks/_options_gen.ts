/** Auto-generated by test_app.py:test_generate_options_js */
export interface OptionsState {
    add_upstream_certs_to_client_chain: boolean
    allow_hosts: string[]
    anticache: boolean
    anticomp: boolean
    block_global: boolean
    block_list: string[]
    block_private: boolean
    body_size_limit: string | undefined
    cert_passphrase: string | undefined
    certs: string[]
    ciphers_client: string | undefined
    ciphers_server: string | undefined
    client_certs: string | undefined
    client_replay: string[]
    client_replay_concurrency: number
    command_history: boolean
    confdir: string
    connect_addr: string | undefined
    connection_strategy: string
    console_focus_follow: boolean
    content_view_lines_cutoff: number
    dns_listen_host: string
    dns_listen_port: number
    dns_mode: string
    dns_server: boolean
    export_preserve_original_ip: boolean
    http2: boolean
    http2_ping_keepalive: number
    ignore_hosts: string[]
    intercept: string | undefined
    intercept_active: boolean
    keep_host_header: boolean
    key_size: number
    listen_host: string
    listen_port: number
    map_local: string[]
    map_remote: string[]
    mode: string
    modify_body: string[]
    modify_headers: string[]
    normalize_outbound_headers: boolean
    onboarding: boolean
    onboarding_host: string
    onboarding_port: number
    proxy_debug: boolean
    proxyauth: string | undefined
    rawtcp: boolean
    readfile_filter: string | undefined
    rfile: string | undefined
    save_stream_file: string | undefined
    save_stream_filter: string | undefined
    scripts: string[]
    server: boolean
    server_replay: string[]
    server_replay_ignore_content: boolean
    server_replay_ignore_host: boolean
    server_replay_ignore_params: string[]
    server_replay_ignore_payload_params: string[]
    server_replay_ignore_port: boolean
    server_replay_kill_extra: boolean
    server_replay_nopop: boolean
    server_replay_refresh: boolean
    server_replay_use_headers: string[]
    showhost: boolean
    ssl_insecure: boolean
    ssl_verify_upstream_trusted_ca: string | undefined
    ssl_verify_upstream_trusted_confdir: string | undefined
    stickyauth: string | undefined
    stickycookie: string | undefined
    stream_large_bodies: string | undefined
    tcp_hosts: string[]
    termlog_verbosity: string
    tls_version_client_max: string
    tls_version_client_min: string
    tls_version_server_max: string
    tls_version_server_min: string
    upstream_auth: string | undefined
    upstream_cert: boolean
    validate_inbound_headers: boolean
    view_filter: string | undefined
    view_order: string
    view_order_reversed: boolean
    web_columns: string[]
    web_debug: boolean
    web_host: string
    web_open_browser: boolean
    web_port: number
    web_static_viewer: string | undefined
    websocket: boolean
}

export type Option = keyof OptionsState

export const defaultState: OptionsState = {
    add_upstream_certs_to_client_chain: false,
    allow_hosts: [],
    anticache: false,
    anticomp: false,
    block_global: true,
    block_list: [],
    block_private: false,
    body_size_limit: undefined,
    cert_passphrase: undefined,
    certs: [],
    ciphers_client: undefined,
    ciphers_server: undefined,
    client_certs: undefined,
    client_replay: [],
    client_replay_concurrency: 1,
    command_history: true,
    confdir: "~/.mitmproxy",
    connect_addr: undefined,
    connection_strategy: "eager",
    console_focus_follow: false,
    content_view_lines_cutoff: 512,
    dns_listen_host: "",
    dns_listen_port: 53,
    dns_mode: "regular",
    dns_server: false,
    export_preserve_original_ip: false,
    http2: true,
    http2_ping_keepalive: 58,
    ignore_hosts: [],
    intercept: undefined,
    intercept_active: false,
    keep_host_header: false,
    key_size: 2048,
    listen_host: "",
    listen_port: 8080,
    map_local: [],
    map_remote: [],
    mode: "regular",
    modify_body: [],
    modify_headers: [],
    normalize_outbound_headers: true,
    onboarding: true,
    onboarding_host: "mitm.it",
    onboarding_port: 80,
    proxy_debug: false,
    proxyauth: undefined,
    rawtcp: true,
    readfile_filter: undefined,
    rfile: undefined,
    save_stream_file: undefined,
    save_stream_filter: undefined,
    scripts: [],
    server: true,
    server_replay: [],
    server_replay_ignore_content: false,
    server_replay_ignore_host: false,
    server_replay_ignore_params: [],
    server_replay_ignore_payload_params: [],
    server_replay_ignore_port: false,
    server_replay_kill_extra: false,
    server_replay_nopop: false,
    server_replay_refresh: true,
    server_replay_use_headers: [],
    showhost: false,
    ssl_insecure: false,
    ssl_verify_upstream_trusted_ca: undefined,
    ssl_verify_upstream_trusted_confdir: undefined,
    stickyauth: undefined,
    stickycookie: undefined,
    stream_large_bodies: undefined,
    tcp_hosts: [],
    termlog_verbosity: "info",
    tls_version_client_max: "UNBOUNDED",
    tls_version_client_min: "TLS1_2",
    tls_version_server_max: "UNBOUNDED",
    tls_version_server_min: "TLS1_2",
    upstream_auth: undefined,
    upstream_cert: true,
    validate_inbound_headers: true,
    view_filter: undefined,
    view_order: "time",
    view_order_reversed: false,
    web_columns: ["tls", "icon", "path", "method", "status", "size", "time"],
    web_debug: false,
    web_host: "127.0.0.1",
    web_open_browser: true,
    web_port: 8081,
    web_static_viewer: "",
    websocket: true,
}
