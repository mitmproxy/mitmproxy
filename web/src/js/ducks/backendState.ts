/**
 * serverState houses properties about the current mitmproxy instance that are not options,
 * e.g. the list of available content views or the current version.
 */

export const RECEIVE = 'STATE_RECEIVE'
export const UPDATE = 'STATE_UPDATE'

export interface ServerInfo {
    description: string
    full_spec: string
    is_running: boolean
    last_exception: string | null
    listen_addrs: [string, number][]
    type: string
    wireguard_conf?: string
}

export interface BackendState {
    available: boolean
    version: string
    contentViews: string[]
    servers: ServerInfo[]
}

export const defaultState: BackendState = {
    available: false,
    version: "",
    contentViews: [],
    servers: [],
};

export default function reducer(state = defaultState, action): BackendState {
    switch (action.type) {
        case RECEIVE:
        case UPDATE:
            return {
                ...state,
                available: true,
                ...action.data
            }
        default:
            return state
    }
}
