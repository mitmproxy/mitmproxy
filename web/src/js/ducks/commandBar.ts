export const TOGGLE_VISIBILITY = "COMMANDBAR_TOGGLE_VISIBILITY";

interface CommandBarState {
    visible: boolean;
}

export const defaultState: CommandBarState = {
    visible: false,
};

export default function reducer(state = defaultState, action): CommandBarState {
    switch (action.type) {
        case TOGGLE_VISIBILITY:
            return {
                ...state,
                visible: !state.visible,
            };

        default:
            return state;
    }
}

export function toggleVisibility() {
    return { type: TOGGLE_VISIBILITY };
}
