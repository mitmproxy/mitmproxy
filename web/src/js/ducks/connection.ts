export enum ConnectionState {
    INIT = "CONNECTION_INIT",
    FETCHING = "CONNECTION_FETCHING", // WebSocket is established, but still fetching resources.
    ESTABLISHED = "CONNECTION_ESTABLISHED",
    ERROR = "CONNECTION_ERROR",
    OFFLINE = "CONNECTION_OFFLINE", // indicates that there is no live (websocket) backend.
}

interface ConnState {
    state: ConnectionState;
    message?: string;
}

const defaultState: ConnState = {
    state: ConnectionState.INIT,
    message: undefined,
};

export default function reducer(state = defaultState, action): ConnState {
    switch (action.type) {
        case ConnectionState.ESTABLISHED:
        case ConnectionState.FETCHING:
        case ConnectionState.ERROR:
        case ConnectionState.OFFLINE:
            return {
                state: action.type,
                message: action.message,
            };

        default:
            return state;
    }
}

export function startFetching() {
    return { type: ConnectionState.FETCHING };
}

export function connectionEstablished() {
    return { type: ConnectionState.ESTABLISHED };
}

export function connectionError(message) {
    return { type: ConnectionState.ERROR, message };
}

export function setOffline() {
    return { type: ConnectionState.OFFLINE };
}
