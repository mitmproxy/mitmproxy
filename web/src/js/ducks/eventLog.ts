import * as store from "./utils/store";

export const ADD = "EVENTS_ADD";
export const RECEIVE = "EVENTS_RECEIVE";
export const TOGGLE_VISIBILITY = "EVENTS_TOGGLE_VISIBILITY";
export const TOGGLE_FILTER = "EVENTS_TOGGLE_FILTER";

type LogLevel = "debug" | "info" | "web" | "warn" | "error";

export interface EventLogItem extends store.Item {
    message: string;
    level: LogLevel;
}

interface EventLogState extends store.State<EventLogItem> {
    visible: boolean;
    filters: { [level in LogLevel]: boolean };
}

const defaultState: EventLogState = {
    visible: false,
    filters: { debug: false, info: true, web: true, warn: true, error: true },
    ...store.defaultState,
};

export default function reduce(
    state: EventLogState = defaultState,
    action
): EventLogState {
    switch (action.type) {
        case TOGGLE_VISIBILITY:
            return {
                ...state,
                visible: !state.visible,
            };

        case TOGGLE_FILTER:
            const filters = {
                ...state.filters,
                [action.filter]: !state.filters[action.filter],
            };
            return {
                ...state,
                filters,
                ...store.reduce(
                    state,
                    store.setFilter<EventLogItem>((log) => filters[log.level])
                ),
            };

        case ADD:
        case RECEIVE:
            return {
                ...state,
                ...store.reduce(
                    state,
                    store[action.cmd](
                        action.data,
                        (log) => state.filters[log.level]
                    )
                ),
            };

        default:
            return state;
    }
}

export function toggleFilter(filter: LogLevel) {
    return { type: TOGGLE_FILTER, filter };
}

export function toggleVisibility() {
    return { type: TOGGLE_VISIBILITY };
}

export function add(message: string, level: LogLevel = "web") {
    let data = {
        id: Math.random().toString(),
        message,
        level,
    };
    return {
        type: ADD,
        cmd: "add",
        data,
    };
}
