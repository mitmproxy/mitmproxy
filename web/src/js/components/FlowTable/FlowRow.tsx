import React, { useCallback } from "react";
import classnames from "classnames";
import { Flow } from "../../flow";
import { useAppDispatch, useAppSelector } from "../../ducks";
import { select } from "../../ducks/flows";
import * as columns from "./FlowColumns";

type FlowRowProps = {
    flow: Flow;
    selected: boolean;
    highlighted: boolean;
};

export default React.memo(function FlowRow({
    flow,
    selected,
    highlighted,
}: FlowRowProps) {
    const dispatch = useAppDispatch(),
        displayColumnNames = useAppSelector(
            (state) => state.options.web_columns
        ),
        className = classnames({
            selected: selected,
            highlighted: highlighted,
            intercepted: flow.intercepted,
            "has-request": flow.type === "http" && flow.request,
            "has-response": flow.type === "http" && flow.response,
        });

    const onClick = useCallback(
        (e) => {
            // a bit of a hack to disable row selection for quickactions.
            let node = e.target;
            while (node.parentNode) {
                if (node.classList.contains("col-quickactions")) return;
                node = node.parentNode;
            }
            dispatch(select(flow.id));
        },
        [flow]
    );

    const displayColumns = displayColumnNames
        .map((x) => columns[x])
        .filter((x) => x)
        .concat(columns.quickactions);

    return (
        <tr className={className} onClick={onClick}>
            {displayColumns.map((Column) => (
                <Column key={Column.name} flow={flow} />
            ))}
        </tr>
    );
});
