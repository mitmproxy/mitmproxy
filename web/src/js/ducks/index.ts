import {
    AnyAction,
    applyMiddleware,
    combineReducers,
    compose,
    createStore as createReduxStore,
    PreloadedState,
    Store
} from "redux"
import eventLog from "./eventLog"
import flows from "./flows"
import ui from "./ui/index"
import connection from "./connection"
import options from './options'
import commandBar from "./commandBar";
import thunk, {ThunkAction, ThunkDispatch, ThunkMiddleware} from "redux-thunk";
import {TypedUseSelectorHook, useDispatch, useSelector} from "react-redux";
import backendState from "./backendState";
import options_meta from "./options_meta";


// @ts-ignore
const composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;

export const rootReducer = combineReducers({
    commandBar,
    eventLog,
    flows,
    connection,
    ui,
    options,
    options_meta,
    backendState,
});
export type RootState = ReturnType<typeof rootReducer>


export type AppThunk<ReturnType = void> = ThunkAction<ReturnType,
    RootState,
    unknown,
    AnyAction>

export const createAppStore = (preloadedState?: PreloadedState<RootState>): Store<RootState> => {
    return createReduxStore(
        rootReducer,
        preloadedState,
        composeEnhancers(applyMiddleware(
            thunk as ThunkMiddleware<RootState>
        )))
};

export const store = createAppStore(undefined);

// this would be correct, but PyCharm bails on it
// export type AppDispatch = typeof store.dispatch
// instead:
export type AppDispatch = ThunkDispatch<RootState, void, AnyAction>
export const useAppDispatch = () => useDispatch<AppDispatch>()
export const useAppSelector: TypedUseSelectorHook<RootState> = useSelector
