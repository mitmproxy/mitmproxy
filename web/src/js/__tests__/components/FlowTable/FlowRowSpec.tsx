import * as React from "react";
import FlowRow from "../../../components/FlowTable/FlowRow";
import { fireEvent, render, screen } from "../../test-utils";
import { TStore } from "../../ducks/tutils";

test("FlowRow", async () => {
    const store = TStore();
    const tflow = store.getState().flows.list[3];
    const { asFragment } = render(
        <table>
            <tbody>
                <FlowRow flow={tflow} selected highlighted />
            </tbody>
        </table>,
        { store },
    );
    expect(asFragment()).toMatchSnapshot();

    expect(store.getState().flows.selected[0]).not.toBe(tflow.id);
    fireEvent.click(screen.getByText("QUERY"));
    expect(store.getState().flows.selected[0]).toBe(tflow.id);
});
