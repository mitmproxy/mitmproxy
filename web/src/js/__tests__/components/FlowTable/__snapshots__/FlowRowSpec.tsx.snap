// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FlowRow 1`] = `
<DocumentFragment>
  <table>
    <tbody>
      <tr
        class="selected highlighted intercepted has-request has-response"
      >
        <td
          class="col-tls col-tls-https"
        />
        <td
          class="col-icon"
        >
          <div
            class="resource-icon resource-icon-websocket"
          />
        </td>
        <td
          class="col-path"
        >
          <i
            class="fa fa-fw fa-pause pull-right"
          />
          <i
            class="fa fa-fw fa-exclamation pull-right"
          />
          <span
            class="marker pull-right"
          />
          http://address:22/path
        </td>
        <td
          class="col-method"
        >
          WSS
        </td>
        <td
          class="col-status"
          style="color: darkgreen;"
        >
          200
        </td>
        <td
          class="col-size"
        >
          43b
        </td>
        <td
          class="col-time"
        >
          5s
        </td>
        <td
          class="col-quickactions"
        >
          <div>
            <a
              class="quickaction"
              href="#"
            >
              <i
                class="fa fa-fw fa-play text-success"
              />
            </a>
          </div>
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;
